trio/_core/tests/test_io.py

```diff
@@ -352,3 +351,93 @@ def check(*, expected_readers, expected_writers):
+            nursery.cancel_scope.cancel()
+
+    # We hit different paths on Windows depending on whether we close the last
+    # handle to the object (which produces a LOCAL_CLOSE notification and
+    # wakes up wait_readable), or only close one of the handles (which leaves
+    # wait_readable pending until cancelled).
+    with stdlib_socket.socket() as s, s.dup() as s2:  # noqa: F841
+        async with trio.open_nursery() as nursery:
+            nursery.start_soon(allow_OSError, trio.hazmat.wait_readable, s)
+            await wait_all_tasks_blocked()
+            s.close()
+            await wait_all_tasks_blocked()
+            nursery.cancel_scope.cancel()
+
+    # A more elaborate case, with two tasks waiting. On windows and epoll,
+    # the two tasks get muxed together onto a single underlying wait
+    # operation. So when they're cancelled, there's a brief moment where one
+    # of the tasks is cancelled but the other isn't, so we try to re-issue the
+    # underlying wait operation. But here, the handle we were going to use to
+    # do that has been pulled out from under our feet... so test that we can
+    # survive this.
+    a, b = stdlib_socket.socketpair()
+    with a, b, a.dup() as a2:  # noqa: F841
+        a.setblocking(False)
+        b.setblocking(False)
+        fill_socket(a)
+        async with trio.open_nursery() as nursery:
+            nursery.start_soon(allow_OSError, trio.hazmat.wait_readable, a)
+            nursery.start_soon(allow_OSError, trio.hazmat.wait_writable, a)
+            await wait_all_tasks_blocked()
+            a.close()
+            nursery.cancel_scope.cancel()
+
+    # A similar case, but now the single-task-wakeup happens due to I/O
+    # arriving, not a cancellation, so the operation gets re-issued from
+    # handle_io context rather than abort context.
+    a, b = stdlib_socket.socketpair()
+    with a, b, a.dup() as a2:  # noqa: F841
+        print("a={}, b={}, a2={}".format(a.fileno(), b.fileno(), a2.fileno()))
+        a.setblocking(False)
+        b.setblocking(False)
+        fill_socket(a)
+        e = trio.Event()
+
+        # We want to wait for the kernel to process the wakeup on 'a', if any.
+        # But depending on the platform, we might not get a wakeup on 'a'. So
+        # we put one task to sleep waiting on 'a', and we put a second task to
+        # sleep waiting on 'a2', with the idea that the 'a2' notification will
+        # definitely arrive, and when it does then we can assume that whatever
+        # notification was going to arrive for 'a' has also arrived.
+        async def wait_readable_a2_then_set():
+            await trio.hazmat.wait_readable(a2)
+            e.set()
+
+        async with trio.open_nursery() as nursery:
+            nursery.start_soon(allow_OSError, trio.hazmat.wait_readable, a)
+            nursery.start_soon(allow_OSError, trio.hazmat.wait_writable, a)
+            nursery.start_soon(allow_OSError, wait_readable_a2_then_set)
```

Comment:
You're not closing a2, so I don't think this needs allow_OSError?

Reply:
Good catch :-)